tests/e2e/plugins/proxy-auth.php

```diff
@@ -37,6 +37,12 @@ function () {
 add_filter(
 	'get_user_option_googlesitekit_auth_scopes',
 	function () {
+		global $wp_current_filter;
+
+		if ( count( array_keys( $wp_current_filter, 'get_user_option_googlesitekit_auth_scopes', true ) ) > 1 ) {
+			return null;
+		}
```

Comment:
Kudos for finding a minimal solution to this! I'm slightly concerned however as to the ability for us to understand what this does in the future and why it's here as it's not very intuitive. I've never actually seen `array_keys` used with the second argument 😄 

At a minimum, let's add an explanation as to what this does and why.

As part of this, I spent some time looking into an alternative which I wasn't able to find something quite as clean. I was able to find something else that worked and is a bit more explicit. Here's a test I added to the `Analytics_4Test` to test it, but essentially the function unhooks itself before calling the method that would otherwise result in recursion before rehooking itself which is something you do see from time to time. It's a little odd to make it work in this context since we need a reference to the function itself. You can imagine it would be cleaner using a global function reference like `_force_all_scopes` – no `$ref` or closure oddities would be needed.

```php
	public function test_recursion() {
		remove_all_filters( 'googlesitekit_auth_scopes' );
		remove_all_filters( 'get_user_option_googlesitekit_auth_scopes' );
		$this->analytics->register();
		// Fake all scopes have been granted.
		$ref     = new \stdClass;
		$ref->fn = function () use ( $ref ) {
			remove_filter( 'get_user_option_googlesitekit_auth_scopes', $ref->fn );
			$required_scopes = ( new OAuth_Client( Plugin::instance()->context() ) )->get_required_scopes();
			add_filter( 'get_user_option_googlesitekit_auth_scopes', $ref->fn );

			return $required_scopes;
		};
		add_filter( 'get_user_option_googlesitekit_auth_scopes', $ref->fn );
		$this->authentication->get_oauth_client()->has_sufficient_scopes();
	}
```

I'm happy to keep things as you have it so long as it's a bit more clear what's going on to save someone the source diving in the future 😄

Reply:
Hey @aaemnnosttv, thanks for the feedback here. To be honest, I'd not considered this alternative solution where we unhook the filter before the call to `get_required_scopes()` and then restore it after, although like so many good ideas it seems quite obvious now you've said it.

I must admit I'm somewhat torn between the two - on the one hand my initial solution provides a touch more control over the flow with it explicitly determining what the return value is on the recursed invocation. However by the same token it's arguably a bit more brittle as it implies some knowledge of the caller code at that point, i.e. that it will handle `null` correctly.

On the other hand your proposal leaves the caller code to run essentially unmodified beyond that initial call. This feels a bit more flexible, and less tied to the current implementation. Of course the fact is we still _are_ modifying things by virtue of the initial override so it's somewhat fortunate that it continues to work once the hook is subsequently removed, but the fact is that it does.

Weighing this up, alongside the more explicit nature of the code in your version, on balance I think we should take your approach - hopefully it's more resilient to future changes in the caller code.

I have therefore updated `proxy-auth.php` and `gcp-auth.php` to take this approach instead - also using a global variable as suggested.

Noting you did mention you're happy to keep things as they are, but also posted this alternative suggestion, I hope you are also happy to take this approach instead. :)